ZenScript/src/main/java/stanhebben/zenscript/expression/ExpressionMapIndexSet.java

```diff
@@ -38,8 +38,8 @@ public ZenType getType() {
     public void compile(boolean result, IEnvironmentMethod environment) {
         if(result) {
             map.compile(result, environment);
-            index.compile(result, environment);
-            value.compile(result, environment);
+            index.cast(getPosition(), environment, ((ZenTypeAssociative)map.getType()).getKeyType()).compile(result, environment);
```

Comment:
cast the key/value before compiling(needed for them to work properly, especially with boxing/Strings)